pandas/core/ops.py

```diff
@@ -709,6 +709,15 @@ def wrapper(left, right, name=name, na_op=na_op):
 
         left, right = _align_method_SERIES(left, right)
 
+        if is_timedelta64_dtype(left) and isinstance(right, pd.DatetimeIndex):
```

Comment:
don't special case things *here*, this is the point of the _TimeOp class.

Reply:
This is the appropriate place.  Adding additional wrapping/unwrapping in _TimeOp obscures whats going on.  Even if it wasn't a third(!) level of wrapping, b/c of namespacing/closure issues it is essentially impossible to handle overflow checks correctly with the current setup.

Eventually this is going to _have_ to look like:

```
def wrapper(...)
    if isinstance(right, ABCDataFrame):
        return NotImplemented
    elif is_timedelta64_dtype(left):
         [dispatch to TimedeltaIndex op]
    elif (is_datetime64_dtype(left) or is_datetime64tz_dtype(left)):
         [dispatch to DatetimeIndex op]

    [everything else]
```

Everything _TimeOp does is done (better) in the index classes.  Anything other than dispatching to those methods is unnecessary duplication and an invitation to inconsistency.

Reply:
so let's untangle that first. This bug fix just addsmore things to undo later.

Reply:
> This bug fix just addsmore things to undo later.

This won't need to be undone.  Eventually the `and isinstance(right, pd.DatetimeIndex)` part of `if is_timedelta64_dtype(left) and isinstance(right, pd.DatetimeIndex):` is removed and the timedelta64_dtype case is complete.

I've given the order of edits quite a bit of thought.  Transitioning towards the dispatch approach case-by-case and implementing corresponding tests along the way is the way to go.

> so let's untangle that first.

Untangle which first?  The mess of closures that prevents overflow checks?

I guess if you want to untangle _TimeOp independently, #18832 is a step in that direction, is orthogonal to the other outstanding PRs.

Reply:
well, until #18832 this needs to change as I have indicated.

Reply:
#18832 is orthogonal.

What change have you indicated? Not clear on what "untangle this first" refers to.

Reply:
well I don't want this here. put it where the other conversions are. that can be reactored at some point if its worthile.

Reply:
I guess if I squint it kind of makes sense to avoid fixing the overflow-check bug in this PR and instead do it separately.  Will change.